File: Python/shortest_path.py
def bfs(graph, source, destination):
    path_list = [[source]]
    path_index = 0
    
    previous_nodes = {source}
    
    if source == destination:
        return path_list[0]
    
    while path_index < len(path_list):
        current_path = path_list[path_index]
        last_node = current_path[-1]
        next_nodes = graph[last_node]
        
        if destination in next_nodes:
            current_path.append(destination)
            return current_path
        
        for next_node in next_nodes:
            if not next_node in previous_nodes:
                new_path = current_path[:]
                new_path.append(next_node)
                path_list.append(new_path)
                previous_nodes.add(next_node)
        path_index += 1
    return []          

def available_moves(idx_to_check, maze):
    row_idx, el_idx = idx_to_check
    directions_available = []
    
    if row_idx - 1 >= 0 and maze[row_idx - 1][el_idx] != 'W': # check north
        directions_available.append("N")
    if row_idx + 1 < len(maze) and maze[row_idx + 1][el_idx] != 'W': # check south
        directions_available.append("S")
    if el_idx - 1 >= 0 and maze[row_idx][el_idx - 1] != 'W': # check west
        directions_available.append("W")
    if el_idx + 1 < len(maze[row_idx]) and maze[row_idx][el_idx + 1] != 'W': # check east
        directions_available.append("E")
    return directions_available

def build_graph(maze):
    nodes = []
    graph = {}
    
    for row in range(len(maze)):
        for col in range(len(maze)):
            nodes.append((row,col))

    for node in nodes:
        connecting_nodes = []
        for dir in available_moves(node, maze):
            if dir == 'N':
                r,c = node
                r -= 1
                connecting_nodes.append((r,c))
            if dir == 'S':
                r,c = node
                r += 1
                connecting_nodes.append((r,c))
            if dir == 'W':
                r,c = node
                c -= 1
                connecting_nodes.append((r,c))
            if dir == 'E':
                r,c = node
                c += 1
                connecting_nodes.append((r,c))
        graph[node] = connecting_nodes
    return graph

def path_finder(maze):
    maze_split = maze.split("\n")
    maze_traveled = [[] for el in range(len(maze_split))]
    N = len(maze_split)

    N = len(maze_split) 
    last_node = (N - 1, N - 1)
    first_node = (0,0)
    
    visited = bfs(build_graph(maze_split), first_node, last_node)
    print("visited is ", visited)
    
    if len(visited) == 0:
        return False
    else:
        return len(visited) - 1

File: CPP/shortest_path.cpp
#include <iostream>
#include <map>
#include <vector>
#include <string>
#include <set>

class Node{
public:
    int row;
    int col;

    Node(){}

    Node(int r, int c) {
      row = r;
      col = c;
    }

    void set(int r, int c) {
      row = r;
      col = c;
    }

    bool in_vector(std::vector<Node> v) {
      return std::find(v.begin(), v.end(), (*this)) != v.end();
    }

    bool in_set(std::set<Node> s) {
      return std::find(s.begin(), s.end(), (*this)) != s.end();
    }

    bool operator<(const Node& other) const {
    if (row < other.row) {
        return true;
    } else if (row == other.row) {
        return col < other.col;
    } else {
        return false;
    }
  }

    bool operator==(const Node& other) const {
      if(row == other.row && col == other.col) {
        return true;
      } else {
        return false;
      }
    }
};

std::map<Node, std::vector<Node>> build_graph(std::vector<std::string> maze) {
  std::map<Node, std::vector<Node>> graph;

  for(int row = 0; row < maze.size(); row++) {
    for(int col = 0; col < maze[row].length(); col++) {
      Node node(row, col);
      std::vector<Node> connecting_nodes;
        if(row - 1 >= 0 && maze[row - 1][col] != 'W') {
          connecting_nodes.push_back(Node(row -1, col));
        }
        if(row + 1 < maze.size() && maze[row + 1][col] != 'W') {
          connecting_nodes.push_back(Node(row + 1, col));
        }
        if(col - 1 >= 0 && maze[row][col - 1] != 'W') {
          connecting_nodes.push_back(Node(row, col - 1));
        }
        if(col + 1 < maze[row].length() && maze[row][col + 1] != 'W') {
          connecting_nodes.push_back(Node(row, col + 1));
        }
        graph[node] = connecting_nodes;
    }
  }
  
  return graph;
}

std::vector<Node> bfs(std::map<Node, std::vector<Node>> graph, Node source, Node destination) {
  std::vector<std::vector<Node>> path_list = {{source}};
  int path_index = 0;

  std::set<Node> previous_nodes;

  if (source == destination) {
    return path_list[0];
  }

  while (path_index < path_list.size()) {
    std::vector<Node> current_path = path_list[path_index];
    Node last_node = current_path.back();
    std::vector<Node> next_nodes = graph.at(last_node);

    if (destination.in_vector(next_nodes)) {
      current_path.push_back(destination);
      return current_path;
    }

    for(Node next_node : next_nodes) {
      if(!next_node.in_set(previous_nodes)) {
        std::vector<Node> new_path = current_path;
        new_path.push_back(next_node);
        path_list.push_back(new_path);
        previous_nodes.insert(next_node);
      }
    }
    path_index++;
  }

  std::vector<Node> empty;
  return empty;
}

std::vector<std::string> split_string(std::string map_string) {
  std::vector<std::string> split_string;
  int start_of_line = 0;
  std::size_t next_newline = map_string.find("\n");

  while(next_newline != std::string::npos) {
    split_string.push_back(map_string.substr(start_of_line, next_newline - start_of_line));
    start_of_line = next_newline + 1;
    next_newline = map_string.find("\n", start_of_line);
  }

  if (start_of_line < map_string.length()) {
       split_string.push_back(map_string.substr(start_of_line));
   }

  return split_string;
}

bool path_finder(std::string maze) {
  std::vector<std::string> map = split_string(maze);
  std::map<Node, std::vector<Node>> graph = build_graph(map);
  int map_dimension = map.size();
  Node start(0,0);
  Node end(map_dimension - 1, map_dimension - 1); // n * n maze

  std::vector<Node> path = bfs(graph, start, end);
  return path.size() != 0;
}
